### ISSUE TITLE
Allow log line max size to be configurable

### ISSUE THREAD

**Issue by someword** [moby/moby#32923]
Prior to docker 1.13 containers could emit large log lines and docker passed the lines along intact to the given log-driver destination (fluentd, journald, json-file, etc).   I'm sure some upper bound existed but in my quick testing on 1.12.6 I was able to run a container which printed to STDOUT a 65K JSON object that got logged in tact as 1 line.   In my shop for better or worse log lines of 16k - 64k are being written by applications.     

As of docker 1.13 log lines are limited to 16k per log line.   Lines longer than this are split into multiple log entries which makes it difficult to view specific events in a log aggregation tool.   If people are writing log messages in a format like JSON then those log entries are not able to be parsed by down stream log aggregation tools because the JSON is now invalid as single objects got split into multiple log lines.

I would like to be able to override this (https://github.com/moby/moby/blob/master/daemon/logger/copier.go#L13 ) constant which I believe is setting the upper bound on the log size.

A log-driver option like --log-opt max-log-line=64k would allow users to choose if they want to emulate the behavior in previous versions of docker by setting a larger max than the default of 16k.

**Comment by dperny** [comment]
2 things: 

First:

> for better or worse log lines of 16k - 64k are being written by applications

This is beautiful in a perversely sublime way, like the rolling smoke of heavy machinery, thick and untempered, from a bygone era before environmental regulations.

Second: I'm going to be more surprised if log lines that big DON'T cause issues with `service logs` if you try to get them through the docker CLI. That's REALLY big and I have no idea what will happen if we try to send them over the wire. I can only assume that it will likely crash at least the CLI, and beyond that probably eat your dog and attract the incomprehensible maddening gaze of the Great Old Ones. 

If we do support something like this, modifications to handle it (like chunking log messages and reconstituting them on the CLI) will likely be needed. Keep me posted on the progression of this issue.

**Comment by someword** [comment]
@dperny I was worried about my dog being eaten as well!  So i got another for HA!

    However in the versions before docker 1.13 we read logs via 'docker logs' and or indirectly through 'kubectl logs' and can read log lines that are 65K without any noticeable impact to any players.   Obviously though it's maddening to read logs of this magnitude through the cli:)  

So i have a patch that add the log option --max-log-size which makes the max log line configurable for folks like me that are stuck with apps that write log lines which contain things like oh say the entire linux kernel source code in one log line ... not really that exactly but you get the idea.

I've not run the full test suite against it yet but I have manually tested.  The branch is here https://github.com/someword/moby/tree/32923-max-log-line.  I'm certain it will need some further twiddling.   However when you say things like 'chunking log messages' i get nervous that my rudimentary patch is possibly a little naive.   I'm curious if the approach of just restoring the ability to have the previous docker behavior by upping the buffer size not going to be an acceptable approach?    I had planned on running the full test suite on my patch, fixing the docs and submitting the PR.   But if the approach is not acceptable ...

Thanks for the feedback!

**Comment by dperny** [comment]
I might be wrong as well. I usually get to work a couple of layers removed from the parts that I think a larger message size is going to cause issues with, so my knowledge of it is really limited.

**Comment by someword** [comment]
So I was expecting the engine documentation to be in this same repo (moby/moby).   But I could not figure out where it is:( So I (incorrectly?) created a corresponding documentation branch here https://github.com/someword/docker.github.io/tree/32923-max-log-line-docs.    Apologies in advance if that is not what I should be doing for the docs.

**Comment by richm** [comment]
+1 for configurable logging size

**Comment by cpuguy83** [comment]
Cross-posting from the PR for this:


> I'm not sure why we'd need to support this.
We already support partial log lines, the other end just needs to understand that it's a partial message rather than filling up the memory on the docker instance.

>Setting a different value is either going to be "set a really big value that we'll never ever exceed" or "oh crap, we need a little more... and a little more... little more". Either way it makes the splitting useless.
Where as supporting the partial message takes care of this.

**Comment by richm** [comment]
> tl;dr there will always be a case where a logging service will need to reconstruct a partial message

Sure. But we're talking about a 99% case here - in most cases, a user can set a large value and be assured that their apps are not going to exceed the limit, and still be well within the container memory limits.

The current limit 16k is purely arbitrary anyway, afaict.

**Comment by cpuguy83** [comment]
@richm "arbitrary" in that it was a nice number that seemed like plenty for basically anyone.

**Comment by portante** [comment]
FWIW, `stdout` and `stderr` streams, for better or for worse, are being split by `newline` characters to form individual payloads we interpret as a "log message".   We _**always**_ need an upper bound on how big those payloads can grow to keep memory usage bounded when processing the stream.  Adding the metadata to indicate messages split at the boundary will always be necessary.  Therefore, tools that need or want to reconstruct the original byte stream are required to utilize that metadata.

Adding a configuration parameter seems to add additional complexity to docker in order to alleviate perceived complexity supporting the method of stream splitting.  Users would have to understand how this parameter must be sized in relation to the `max-buffer-size`.  Documenting the interactions would be required, and take effort to get right.

(Does Docker handle well when a max-buffer-size is lower than 16K today?)

The fact that a stream of bytes is being interpreted as a series of JSON documents delimited by newline characters, or XML, or any other scheme, does not appear to be material, as docker is not the entity doing the interpreting.  As such, it would seem a system must be ready to handle the method of how that stream is split by docker before interpreting.

It seems then that docker is only responsible for ensuring that systems can reconstruct the original byte stream using the metadata provided.

**Comment by cpuguy83** [comment]
Thanks @portante, much better than I could say it.

Closing since this is not something we want to support instead preferring log consumers to handle the case of log messages being broken into multiple parts.

Thanks again to all. 👼

**Comment by richm** [comment]
So we're stuck with an arbitrary 16k log size.
/me isn't looking forward to implementing this joining logic in fluentd, rsyslog, whatever else, for 90% of our customers, rather than the 10% that would have been left over if the other 90% could just adjust the max log size to fit their needs . . .

**Comment by thaJeztah** [comment]
@richm ironically, the change to use arbitrary 16k log size seem to have come from one of your colleages https://github.com/moby/moby/commit/513ec73831269947d38a644c278ce3cac36783b2

**Comment by richm** [comment]
@thaJeztah yes, a person with whom I am long acquainted, and whom I respect a great deal.  There was a good reason to add a limit - the original implementation had no upper bound on the size == oom and other fun.  So we can't have a configurable limit?

**Comment by richm** [comment]
let me ask another way - what if the original fix for https://github.com/moby/moby/commit/513ec73831269947d38a644c278ce3cac36783b2 had included a way to configure the size limit with a -max-log-size parameter or some such?  Would it have been acceptable?

**Comment by thaJeztah** [comment]
t.b.h. even with a configurable limit, there's still situations where the limit is exceeded; not handling the "partial messages" correctly on the receiver's end is then relying on the wrong assumption that _every_  log entry fits in the limit that was set. I don't think that's the right thing to do.

**Comment by cpuguy83** [comment]
@richm The trouble is, no matter what the splitting needs to be supported on the other end. Changing in Docker gives the memory management problem to docker. Meanwhile in Docker we just want to get the log out of Docker as quickly as possible.

Supporting a variable size buffer also means far greater complexity in the log copier.
So you are asking Docker to take on greater complexity without any real win since the logging endpoint still needs to support re-assembling log messages.

This can also be handled in a logging plugin (included in 17.05) where the plugin can re-assemble the message before passing to whatever logging service.

**Comment by richm** [comment]
Yes, I realize this: _no matter what, a log processor (e.g. fluentd) must be able to reconstruct a message from partial fragments_.  I'm just saying that the need for this in many cases can be obviated by allowing a configurable max log message size.  Yes, it shifts the memory burden to Docker.  Yes, it allows administrators the choice, of whether to move that memory management burden to Docker, or move the memory buffering and log stitching complexity/cpu processing to another process.  Administrators could then weigh the costs and figure out what works best for them in their environments.

I didn't think the proposed PR added much additional complexity, but then, I'm not the person who has to maintain it.

I'll take a look at the logging plugin.

I'm done flogging this particular dead horse.

**Comment by jiahaohe** [comment]
When I tested this, I was not able to see the "partial log message metadata" in fluentd. Is this being passed by the fluentd logging driver?

**Comment by someword** [comment]
@jiahaohe only the journald log driver passes through the partial attribute.  All of the other log drivers need to be patched to make this information available.

**Comment by crassirostris** [comment]
@fgrzadkowski

**Comment by bamb00** [comment]
Hi, Is this feature available in version docker 18.09.2?

**Comment by bobzoller** [comment]
I went down this rabbit hole and to save others some time: AFAICT _as of today_ on master, log drivers can implement the [`SizedLogger`](https://github.com/moby/moby/blob/8c0d7e6d52ea04a1523944dda8c6e168dc785ae2/daemon/logger/logger.go#L81-L86) interface, which _will_ override the default 16KB limit.  The only log driver that does so is [`cloudwatchlogs`](https://github.com/moby/moby/blob/cd14846d0cde098bb83037d99104db6fadfef039/daemon/logger/awslogs/cloudwatchlogs.go).

**Comment by cpuguy83** [comment]
SizedLogger is an interface specifically to optimize how cloudwatch does buffering.

Partial log messages now include extra metadata that allow you to stitch the message back together more easily.

The issue here is that we cannot have an infinite buffer to store logs lines so it has to be split at some point.
Enabling this to be configurable is effectively the same as an infinite buffer.

**Comment by bobzoller** [comment]
I see.  Does that mean a custom log driver is still subject to the 16KB limit?

Thanks for the hint about [partial log message metadata](https://github.com/moby/moby/blob/cd14846d0cde098bb83037d99104db6fadfef039/daemon/logger/copier.go#L139-L152) though, I had not come across that helpful tidbit!

**Comment by portante** [comment]
Why do we continue to consider logging implementations where the content of the log stream materially influences the behavior of the implementation?

That is, why would we want to interpret the log data stream for newlines during collection, instead of collecting the data without interpretation in such a way that an interpretation can be performed later off the critical path?

See https://github.com/cri-o/cri-o/pull/1605.

**Comment by cpuguy83** [comment]
Yes a custom driver is susceptible to those limits.

Yes I think we should consider a model that allows for pluging into the stream itself rather than breaking into discrete messages in the daemon.
We'd still need to support the existing model, but a new model where we just don't care about the stream except to wire it up to some collection service would be great.

**Comment by cpuguy83** [comment]
Possibly related to that: https://github.com/moby/moby/issues/40517

I would like to move logdrivers out of the daemon. Spending some time thinking about it and speaking with others, it will likely involve more than just moving the drivers themselves and likely copy the all the log handling into a separate bin (not just the driver/drivers).
If we make that bin configurable someone should be able to inject their own behavior.

Ideally the communication there is as efficient as possible.

**Comment by dsx0** [comment]
I commented my solution at: https://github.com/moby/moby/issues/34620#issuecomment-619369707
hope it helps.
